src/libcore/iter/adapters/chain.rs

```diff
@@ -283,6 +283,36 @@ impl<A, B> FusedIterator for Chain<A, B>
           B: FusedIterator<Item=A::Item>,
 {}
 
+#[stable(feature = "chain_exact_size", since = "1.41.0")]
+impl<A, B> ExactSizeIterator for Chain<A, B>
+where
+    A: ExactSizeIterator,
+    B: ExactSizeIterator<Item = A::Item>,
+{
+    /// Returns the exact number of times the iterator will iterate.
+    ///
+    /// # Overflow Behavior
+    ///
+    /// The method does no guarding against overflows, so calling `len` on an
+    /// iterator with more than [`usize::MAX`] elements either produces the
+    /// wrong result or panics. If debug assertions are enabled, a panic is
+    /// guaranteed.
+    ///
+    /// # Panics
+    ///
+    /// This function might panic if the iterator has more than [`usize::MAX`]
+    /// elements.
+    ///
+    /// [`usize::MAX`]: ../../std/usize/constant.MAX.html
+    fn len(&self) -> usize {
+        match self.state {
+            ChainState::Both => self.a.len() + self.b.len(),
```

Comment:
Wouldn't it be better to make sure we *always* panic on overflow here?

Reply:
That's a good point. The existing methods do not guarantee a panic probably because there, the overflowing add is executed very often. But `ExactSizeIterator::len`  is usually not called very often, so we might have a bit of overhead. Though`.len()` feels like an accessor to me, so maybe users do call it a lot?

I have no preference.

Reply:
Personally I'd make all arithmetic checking until there is evidence that that is actually a performance problem. ;)

Reply:
Shall we go with checked here then? It would simplify the docs too.

Reply:
Sure! Will change that soon.